### DISCUSSION TITLE
Fetch messages once in a central place? (getStaticProps)

### DISCUSSION THREAD

**Discussion by timohausmann** [amannn/next-intl#810] [Q&A] [answered]
Hey, I'm using the pages router, everything works great and now I'm looking into optimization.
I noticed that all pages run the same ``import(`../messages/${locale}.json`)`` inside `getStaticProps`, so each generated `page.json` contains the entire messages again and again which seems like overhead. 

Is this really the way to go or can I fetch the translations once in a central place? 
Another approach could be to split the `${locale}.json` per page, but it could be a hassle to get it right.

**Comment by amannn** [comment, marked as answer]
If you use the same messages on every page and have no need for splitting, you can also fetch them via `App.getInitialProps`. Note that this will disable SSG, **except** for if you define a `getStaticProps` function in your pages (even if it just returns an empty object). Hope this helps!

**Reply by timohausmann** [reply]
Hey, thanks for the suggestion! I first tried `App.getInitialProps` with `getStaticProps` {} but on each route change e.g. to `/settings`, a `/_next/data/settings.json` is loaded (`home.json` for `/home` etc.) with all messages. That must be the Automatic Static Optimization – basically like it was before, in my case I try to reduce this redundancy/network load.

without SSG I only get a single global messages js file upon navigation.  
If I am not mistaken this puts me at one initial SSR per visit, a trade I'm willing to take if I encounter no further issues. 
(I'm trying to make everything work offline with as little cache as possible, that's why I jump through these hoops.)
I wonder if app router could do both, SSG and no redundancy.

**Reply by amannn** [reply]
Ah, right—good catch!

> I wonder if app router could do both, SSG and no redundancy.

It does in fact, yep! It also offers the ability to handle i18n in Server Components, possibly not needing to download messages to the Client in the first place (depending on how interactive your app is of course).

**Reply by timohausmann** [reply]
Haha of course, it would be interesting to see the app router way, but I'm so close to the finishing line now, maybe some day I can refactor that too.
